### ISSUE TITLE
Regarding the issue of 2D fiducials visualization

### ISSUE THREAD

**Issue by mygk2023** [eigenvivek/DiffPose#48]
Hello, I greatly admire your work. When visualizing fiducials, I encountered an issue where the 2D fiducials do not match the 2D X-ray image. Below is my code. I used the vision_marker function to visualize the 2D fiducials. Here, I chose id_numbers = 1 and idx = 43 as an example. I found that the image size is (1436, 1436), but the 2D fiducial coordinates exceed the range of 1436. Could you help me identify where the problem is? In the figure, the red dots are the 2D fiducials I visualized. Also, in the figure, are the blue boxes the real fiducials?

I called the method as follows:
`self.true_projected_fiducials = self.project(gt_pose)`

```

class Evaluator:
    def __init__(self, specimen, idx):
        # Save matrices to device
        self.translate = specimen.translate
        self.flip_xz = specimen.flip_xz
        self.intrinsic = specimen.intrinsic
        self.intrinsic_inv = specimen.intrinsic.inverse()

        # Get gt fiducial locations
        self.specimen = specimen
        self.fiducials = specimen.fiducials
        gt_pose = specimen[idx][1]
        self.idx = idx
        self.true_projected_fiducials = self.project(gt_pose)
        

    def vision_marker(self, image, x):
        import matplotlib.pyplot as plt
        x_2d = x.cpu().detach().numpy()
        x_2d = x_2d.squeeze(0)
        x_2d = x_2d[:, :2]

        image = image.squeeze()  # (1436, 1436)
        H, W = image.shape 

        plt.imshow(image, cmap="gray")
        plt.scatter(x_2d[:, 0], x_2d[:, 1], color='red', marker='o', s=50) 
        plt.savefig("marked_image.png", dpi=300)


    def project(self, pose):
        extrinsic = convert_diffdrr_to_deepfluoro(self.specimen, pose)
        x = perspective_projection(extrinsic, self.intrinsic, self.fiducials)
        self.vision_marker(self.specimen[self.idx][0], x)
        x = -self.specimen.focal_len * torch.einsum(
            "ij, bnj -> bni",
            self.intrinsic_inv,
            pad(x, (0, 1), value=1),  # Convert to homogenous coordinates
        )
        # print(x)
        extrinsic = (
            self.flip_xz.inverse().compose(self.translate.inverse()).compose(pose)
        )
        # print(extrinsic.transform_points(x))
        return extrinsic.transform_points(x)

    def __call__(self, pose):
        pred_projected_fiducials = self.project(pose)
        registration_error = (
            (self.true_projected_fiducials - pred_projected_fiducials)
            .norm(dim=-1)
            .mean()
        )
        registration_error *= 0.194  # Pixel spacing is 0.194 mm / pixel isotropic
        return registration_error
```

The detailed coordinates of the computed fiducials (x) are as follows:
tensor([[[1529.6276,  812.6477],
         [ -32.1971,  668.3568],
         [1388.4550, 1251.4653],
         [ 207.6410, 1191.4645],
         [1344.6030,  786.7796],
         [ 426.8202,  720.1949],
         [1160.3257, 1502.3492],
         [ 563.2103, 1480.8383],
         [ 799.7328, 1539.6133],
         [ 738.6487, 1539.1101],
         [ 936.7755, 1455.2964],
         [ 641.0571, 1424.2174],
         [ 763.6991, 1427.3876],
         [ 713.5483, 1425.6681]]])

![Image](https://github.com/user-attachments/assets/aa93b6c6-5f47-4428-91a9-2c9fe351c145)

**Comment by eigenvivek** [comment]
hi @mygk2023 I was also confused by this when I first started working with this dataset. To answer your questions in reverse, the markers you highlighted in blue are **NOT** the fiducial markers in this dataset. The metal BBs in your blue boxes were implanted during the acquisition of these X-rays, but they're used for a completely different purpose that's independent of these registration experiments.

The actual fiducials, which you've shown in red, were manually annotated on the CT. The reason some of them are outside the X-ray is because the structure that they annotate is outside of the field of view for image 43.

The original data repo has a lot more info about this: https://github.com/rg2/DeepFluoroLabeling-IPCAI2020

![Image](https://github.com/user-attachments/assets/547010be-9291-4341-90bc-0a2c61fcd668)

**Comment by mygk2023** [comment]
@eigenvivek Thank you for your prompt reply! I have a confusion here: when calculating the mTRE, the fiducials are first projected onto the 2D plane, and then converted back into 3D coordinates. However, these resulting 3D coordinates seem to lie on the same plane.

My question is: why don't we directly calculate the mean Target Registration Error（mTRE) in 2D image space, rather than converting points back into a 3D plane and calculating mTRE there?

![Image](https://github.com/user-attachments/assets/cf8a5bb1-e7ec-405c-85f3-56e8e541df74)

**Comment by eigenvivek** [comment]
> why don't we directly calculate the mean Target Registration Error（mTRE) in 2D image space

This choice was mostly due to a lack of familiarity with metrics used to evaluate 2D/3D registration on my part 😛

What you're proposing (measuring TRE in the 2D projective plane) is typically referred to as "mean Projective Error". It's great for determining if the estimated locations of fiducials in 2D are in the correct place. However, it doesn't tell you if the estimated depth of the detector is correct. Unprojecting 2D fiducials onto the 3D detector plane, as is done in the DiffPose paper, lets you quantify this depth disparity. Still, computing mTRE *directly in 3D* is probably the most straightforward metric for evaluating 2D/3D registration, in my opinion.

In our latest work ([`xvr`](https://github.com/eigenvivek/xvr/)), we compare these different metrics. In general mTRE in 3D is the most stringent metric, so I think it makes sense to primarily evaluate using that.
- mPE = mean projective error
- mRPE = mean reproejctive error (what's proposed in DiffPose)
- mTRE = mean target registration error
- dGeo = double geodesic distance (the training loss in DiffPose)

![Image](https://github.com/user-attachments/assets/ad8754c6-bd50-4b9c-9e07-43e82c05471d)
